### ISSUE TITLE
Process crash on inject

### ISSUE THREAD

**Issue by Zeprus** [oleavr/android-inject-custom#2]
Hello,

I've compiled your example for armeabi-v7a and arm64, neither works. Whenever I inject the shared object the target process crashes. I've also tried building an example on my own with the same result. Here's the code I used for my own example: https://pastebin.com/623kFV9D
Below are screenshots of my shell (as URLs because they would be huge).
https://user-images.githubusercontent.com/17969238/63078048-8ca2d200-bf3a-11e9-9a8b-33915ae23869.png
https://user-images.githubusercontent.com/17969238/63078049-8d3b6880-bf3a-11e9-9c78-95646c9c4d90.png

**Comment by qq782645210** [comment]
how to build it ,the environment is ?
help me!
i cannot build it in win10 ndk20

**Comment by qq782645210** [comment]
hello , are you here?

**Comment by mosherubin** [comment]
Hi oleavr,

I'm appending my issue to this one by Zeprus as we seem to have the same problem: the "victim" process crashes when the agent.so payload is injected into the "victim" process space.  I would like to add my findings in the hope you can solve the problem.

**The symptom** 

Although `victim` runs successfully, as soon as `inject` runs, the `victim` process terminates with a "segmentation fault".

**Changes made to the original source files**

Makefile:

```
host_arch := aarch64
host_compiler_triplet := armv7a-linux-androideabi21-
...
frida_version := 12.8.13
```

To enable debugging the problem I inserted many "printf" trace statements in both `victim.c` and `inject.c`, hoping to pinpoint the code statement in either of the files that may have caused the problem.

victim.c:

```
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>

int
main (int argc, char * argv[])
{
  pid_t pid = getpid();
  printf ("Victim running with PID %d\n", pid);

  int count = 0;

  while (1)
  {
    int fd;

    printf ("1 (PID: %d - %d) -----------------\n", pid, ++count);	//debug
    fd = open ("/etc/hosts", O_RDONLY);
    printf ("2\n");	//debug
    if (fd != -1)
    {
      printf ("3\n");	//debug
      close (fd);
      printf ("4\n");	//debug
    }

    printf ("5\n");	//debug
    fd = open ("/etc/passwd", O_RDONLY);
    printf ("6\n");	//debug
    if (fd != -1)
    {
      printf ("7\n");	//debug
      close (fd);
      printf ("8\n");	//debug
    }

    printf ("9\n");	//debug
    sleep (1);
    printf ("10\n");	//debug
  }
}
```

inject.c:

```
#include <frida-core.h>
#include <stdio.h>
#include <stdlib.h>

int
main (int argc, char * argv[])
{
  printf ("10\n");	//debug
  int result = 0;
  FridaInjector * injector;
  int pid;
  GError * error;
  guint id;

  if (argc != 2)
    goto bad_usage;

  printf ("20\n");	//debug
  pid = atoi (argv[1]);
  if (pid <= 0)
    goto bad_usage;

  printf ("30\n");	//debug
  frida_init ();

  printf ("40\n");	//debug
  injector = frida_injector_new ();
  printf ("50\n");	//debug

  error = NULL;
  printf ("60\n");	//debug
  id = frida_injector_inject_library_file_sync (injector, pid, "./agent.so", "example_agent_main", "example data", &error);
  printf ("70 id=%d\n", id);	//debug
  if (error != NULL)
  {
    fprintf (stderr, "%s\n", error->message);
    g_error_free (error);

    result = 1;
  }

  printf ("80\n");	//debug
  frida_injector_close_sync (injector);
  printf ("90\n");	//debug
  g_object_unref (injector);
  printf ("100\n");	//debug

  frida_deinit ();
  printf ("110\n");	//debug

  return result;

bad_usage:
  {
    fprintf (stderr, "Usage: %s <pid>\n", argv[0]);
    return 1;
  }
}
```

**Building the binaries**

I was able to run the project 'make' file successfully:

```
/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi21-clang -Wl,-pie -DANDROID -Os -Wall -fPIC -ffunction-sections -fdata-sections -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -I./ext/frida-core inject.c -o bin/inject -L./ext/frida-core -lfrida-core -fuse-ld=gold -Wl,--icf=all -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--fix-cortex-a8

/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-strip --strip-all bin/inject

/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi21-clang -shared -DANDROID -Os -Wall -fPIC -ffunction-sections -fdata-sections -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -I./ext/frida-gum agent.c -o bin/agent.so -L./ext/frida-gum -lfrida-gum -Wl,--version-script,agent.version -fuse-ld=gold -Wl,--icf=all -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--fix-cortex-a8

/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-strip --strip-all bin/agent.so

/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi21-clang -Wl,-pie -DANDROID -Os -Wall -fPIC -ffunction-sections -fdata-sections -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb victim.c -o bin/victim -fuse-ld=gold -Wl,--icf=all -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--fix-cortex-a8

/mnt/c/Users/brown/Downloads/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-strip --strip-all bin/victim
```
When I dumped the ELF file's header, however, it showed it's still 32-bit, rather than 64:

```
brown@DESKTOP-AE2I7DM:/mnt/d/Private/Dev/Open-Source-Projects/android-inject-custom$ readelf -h bin/victim
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           ARM
  Version:                           0x1
  Entry point address:               0x43c
  Start of program headers:          52 (bytes into file)
  Start of section headers:          4596 (bytes into file)
  Flags:                             0x5000200, Version5 EABI, soft-float ABI
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         9
  Size of section headers:           40 (bytes)
  Number of section headers:         25
  Section header string table index: 24
```

I copied the three binaries (victim, inject, agent.so) to `/data/local/tmp/bin` on my Android device.

**Runtime artifacts**

Running `victim` succeeds:

```
zeroflte:/data/local/tmp/bin # ./victim
Victim running with PID 2808
1 (PID: 2808 - 1) -----------------
2
3
4
5
6
9
10
1 (PID: 2808 - 2) -----------------
2
3
4
5
6
9
10
...
1 (PID: 2808 - 9) -----------------
2
3
4
5
6
9
```

In other words, every second `victim` outputs another series of traces of 1-9, sleeps a second, etc.

Now I run `inject`:

```
zeroflte:/data/local/tmp/bin # ./inject 2808
10
20
30
40
50
60
70 id=1
80
90
100
110
zeroflte:/data/local/tmp/bin #
```

Instantaneously, in the victim console, the following lines are added:

```
Segmentation fault
139|zeroflte:/data/local/tmp/bin #
```

Logcat shows the following failure:

```
05-14 13:13:57.935  2808  2857 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 2857 (victim)
05-14 13:13:58.009  2858  2858 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
05-14 13:13:58.011  2858  2858 F DEBUG   : Build fingerprint: 'samsung/zerofltedv/zeroflte:7.0/NRD90M/G920IDVU4FRD3:user/release-keys'
05-14 13:13:58.012  2858  2858 F DEBUG   : Revision: '11'
05-14 13:13:58.012  2858  2858 F DEBUG   : ABI: 'arm'
05-14 13:13:58.013  2858  2858 F DEBUG   : pid: 2808, tid: 2857, name: victim  >>> ./victim <<<
05-14 13:13:58.013  2858  2858 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0
05-14 13:13:58.013  2858  2858 F DEBUG   :     r0 f0eef360  r1 f0c338e0  r2 f0c338e4  r3 f0f8c291
05-14 13:13:58.014  2858  2858 F DEBUG   :     r4 00000000  r5 00000000  r6 f0eef000  r7 00000003
05-14 13:13:58.014  2858  2858 F DEBUG   :     r8 f0eef468  r9 f0eef000  sl f0eee081  fp 00000002
05-14 13:13:58.014  2858  2858 F DEBUG   :     ip 00000000  sp f0c338e0  lr f0eee0d1  pc 00000000  cpsr 400f0010
05-14 13:13:58.018  2858  2858 F DEBUG   :
05-14 13:13:58.018  2858  2858 F DEBUG   : backtrace:
05-14 13:13:58.028  2858  2858 F DEBUG   :     #00 pc 00000000  <unknown>
05-14 13:13:58.029  2858  2858 F DEBUG   :     #01 pc 000000cf  <anonymous:f0eee000>
```

**Device Specs**

Here are my Android device specs:

```
Device name: Galaxy S6
Model number: SM-G920I
Android version: 7.0 (64-bit)
Kernel version: 3.10.61-13327150, Thu Apr 19 18:49:13 KST 2018
```

**Summary**

Launching `inject` causes a SIGSEGV (SEGV_MAPERR) in the `victim` process.  Can the problem be due to a 32 vs. 64-bit issue?

**Comment by oleavr** [comment]
Thanks for the feedback everyone. Thanks to @meme this is now fixed as of a452c80. Cheers!
